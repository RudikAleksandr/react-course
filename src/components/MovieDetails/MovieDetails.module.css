
.root {
  background: linear-gradient(to top,rgba(0, 0, 0, .9) 0,rgba(0, 0, 0, .9) 100%), url('/src/assets/images/netflixMovies.jpg');
  min-height: 400px;
  width: 100%;
}

.netflixLink {
  color: #f65261;
  font-size: 1.5rem;
}

.netflixLink:hover {
  color: #f65261;
}

.movieName {
  font-weight: 400;
  color: white;
  margin-bottom: 0px;
}

.movieRating {
  text-align: center;
  vertical-align: middle;
  line-height: 2.7rem;
  font-weight: 300;
  display: inline-block;
  font-size: 1.5rem;
  width: 50px;
  height: 50px;
  margin-left: 10px;
  color: #80b859;
  border: 1px solid white;
  border-radius: 50%;
}

.movieReleaseYear,
.movieDuration {
  font-size: 1.5rem;
  color: #f65261;
}

.movieDuration {
  margin-left: 40px;
}

.movieOverview {
  margin-top: 15px;
  font-size: 1.12rem;
  color: #bbb;
  display: -webkit-box;
  -webkit-line-clamp: 7;
  -webkit-box-orient: vertical;
  overflow: hidden;
}

.iconSearch {
  cursor: pointer;
}

.movieSlogan {
  font-size: 1.12rem;
  color: #bbb;
  font-weight: normal;
}
