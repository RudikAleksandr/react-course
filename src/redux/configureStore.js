import { configureStore as configureStoreToolkit } from '@reduxjs/toolkit';
import thunkMiddleware from 'redux-thunk';
import moviesReducer from './moviesSlice';

const configureStore = (preloadedState) => (
  configureStoreToolkit({
    preloadedState,
    reducer: {
      movies: moviesReducer,
    },
    middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(thunkMiddleware),
  })
);

export default configureStore;
