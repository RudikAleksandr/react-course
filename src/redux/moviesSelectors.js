
const selectListMovies = (state) => state.movies.moviesOrder.map(id => state.movies.movies[id]);

const selectSearch = (state) => state.movies.search;

const selectActiveMovieGenre = (state) => state.movies.activeMovieGenre;

const selectActiveSortType = (state) => state.movies.activeSortType;

export {
  selectListMovies,
  selectSearch,
  selectActiveMovieGenre,
  selectActiveSortType
};