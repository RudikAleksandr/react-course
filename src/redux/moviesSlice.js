import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import * as moviesAPI from '../api/moviesAPI';
import { groupByKey } from '../utils/grouping';

export const initialState = {
  movies: {},
  moviesOrder: [],
  search: '',
  activeMovieGenre: '',
  activeSortType: 'release_date'
};

export const getListMovies = createAsyncThunk(
  'movies/getListMovies',
  moviesAPI.getListMovies
);

export const getMovie = createAsyncThunk(
  'movies/getMovie',
  moviesAPI.getMovie
);

export const updateMovie = createAsyncThunk(
  'movies/updateMovie',
  moviesAPI.updateMovie
);

export const createMovie = createAsyncThunk(
  'movies/createMovie',
  moviesAPI.createMovie
);

export const deleteMovie = createAsyncThunk(
  'movies/deleteMovie',
  moviesAPI.deleteMovie
);

const moviesSlice = createSlice({
  name: 'movies',
  initialState,
  reducers: {
    setSearch: (state, { payload }) => {
      state.search = payload;
    },
    setActiveMovieGenre: (state, { payload }) => {
      state.activeMovieGenre = payload;
    },
    setActiveSortType: (state, { payload }) => {
      state.activeSortType = payload;
    }
  },
  extraReducers: {
    [getListMovies.fulfilled]: (state, { payload }) => {
      state.movies = groupByKey('id', payload.data);
      state.moviesOrder = payload.data.map((movie) => movie.id);
    },
    [updateMovie.fulfilled]: (state, { payload }) => {
      state.movies[payload.id] = payload;
    },
  },
});

export const {
  setSearch,
  setActiveMovieGenre,
  setActiveSortType
} = moviesSlice.actions;

export default moviesSlice.reducer;