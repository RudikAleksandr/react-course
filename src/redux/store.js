import { configureStore } from '@reduxjs/toolkit';
import moviesReducer from './moviesSlice';
import thunkMiddleware from 'redux-thunk'

const store = configureStore({
  reducer: {
    movies: moviesReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(thunkMiddleware),
});

export default store;