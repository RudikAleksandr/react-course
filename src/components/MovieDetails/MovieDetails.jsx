import React, { useEffect, useState } from 'react';
import { Row, Col, Image } from 'react-bootstrap';
import { Link, useParams } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import cn from './MovieDetails.module.css';
import searchIcon from './search.svg';

import { getMovie } from '../../redux/moviesSlice';

const MovieDetails = () => {
  const { id } = useParams();
  const [movie, setMovie] = useState({});
  const dispatch = useDispatch();

  useEffect(() => {
    document.documentElement.scrollTop = 0;
    dispatch(getMovie(id)).then(({ payload }) => {
      setMovie(payload);
    });
  }, [id, dispatch]);

  return (
    <Col className={cn.root}>
      <Row className="justify-content-between align-items-end ml-5 mr-5 mt-3">
        <Link className={cn.netflixLink} to="/">
          <b>netflix</b>
          roulette
        </Link>
        <Link to="/">
          <Image
            className={cn.iconSearch}
            src={searchIcon}
            alt="Search icon"
          />
        </Link>
      </Row>
      <Row className="justify-content-between align-items-start m-auto w-75">
        <Image
          width="210px"
          className={cn.imageMovie}
          src={movie.poster_path || ''}
          alt="Movie screensaver"
        />
        <Col className="ml-4">
          <h2 className={cn.movieTitle}>
            {movie.title}
            {' '}
            <span className={cn.movieRating}>{movie.vote_average}</span>
          </h2>
          <h3 className={cn.movieSlogan}>{movie.tagline}</h3>
          <span className={cn.movieReleaseYear}>
            {!!movie.release_date && new Date(movie.release_date).getFullYear()}
          </span>
          <span className={cn.movieDuration}>
            {movie.runtime}
            {' '}
            min
          </span>
          <p className={cn.movieOverview}>{movie.overview}</p>
        </Col>
      </Row>
    </Col>
  );
};

export default MovieDetails;
