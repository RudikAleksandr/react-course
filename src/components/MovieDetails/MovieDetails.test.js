
import { render, screen } from '@testing-library/react';
import { MemoryRouter, Route } from 'react-router';
import MovieDetails from './MovieDetails';
import * as reactRedux from 'react-redux';
import * as React from 'react';
import configureStore from 'redux-mock-store'
import { initialState } from '../../redux/moviesSlice';
import thunk from 'redux-thunk'

const getComponent = (store) => (
  <MemoryRouter>
    <Route>
      <reactRedux.Provider store={store}>
        <MovieDetails />
      </reactRedux.Provider>
    </Route>
  </MemoryRouter>
);

describe('MovieDetails component', () => {
  test('render MovieDetails snapshot', () => {
    const store = configureStore([thunk])({ movies: initialState });
    const component = getComponent(store);
    const { asFragment } = render(component);

    expect(asFragment(component)).toMatchSnapshot();
  });

  test('should check movie in document', async () => {
    const store = configureStore([thunk])({ movies: initialState });
    const component = getComponent(store);
    const movie = { release_date: new Date(), overview: 'overview' };
    const mockDispatch = jest.fn(() => Promise.resolve({ payload: movie }));

    jest.spyOn(reactRedux, 'useDispatch').mockReturnValue(mockDispatch);

    render(component);

    expect(await screen.findByText(movie.overview)).toBeInTheDocument();
  });
});
