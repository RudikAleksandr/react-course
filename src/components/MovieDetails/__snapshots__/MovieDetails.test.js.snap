// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails component render MovieDetails snapshot 1`] = `
<DocumentFragment>
  <div
    class="root col"
  >
    <div
      class="justify-content-between align-items-end ml-5 mr-5 mt-3 row"
    >
      <a
        class="netflixLink"
        href="/"
      >
        <b>
          netflix
        </b>
        roulette
      </a>
      <a
        href="/"
      >
        <img
          alt="Search icon"
          class="iconSearch"
          src="search.svg"
        />
      </a>
    </div>
    <div
      class="justify-content-between align-items-start m-auto w-75 row"
    >
      <img
        alt="Movie screensaver"
        class="imageMovie"
        src=""
        width="210px"
      />
      <div
        class="ml-4 col"
      >
        <h2
          class="movieTitle"
        >
           
          <span
            class="movieRating"
          />
        </h2>
        <h3
          class="movieSlogan"
        />
        <span
          class="movieReleaseYear"
        />
        <span
          class="movieDuration"
        >
           min
        </span>
        <p
          class="movieOverview"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
